### PR TITLE
feat: integrate kyverno

### PR DESCRIPTION
**Pull request by KhaledEmaraDev**
This pull request introduces a new feature that integrates the Kyverno provider with Lula. The Kyverno provider enables Lula to evaluate the domain in the target against a kyverno-json policy.

## Key features of the Kyverno provider integration:

1. **Validation Payload:**

The validation is performed using a payload that specifies the provider (kyverno), domain (kubernetes), and the necessary configuration.
The payload includes the resource rules, Kyverno policy definition, and optional output configuration.

2. **Multiple Policy Support:**

The Kyverno provider allows defining multiple policies within the payload.
Each policy is defined using the Kyverno policy structure, including metadata, rules, and assertions.

3. **Validation Output Control:**

The `output.validation` field in the payload enables controlling which (Policy, Rule) pair determines the validation allowance or denial.
It accepts a comma-separated list of rules in the format policy-name.rule-name.

4. **Observation Output:**

The `output.observations` field in the payload allows specifying (Policy, Rule) pairs to observe violations.
Observed violations will be included in the remarks section of the relevant-evidence in the assessment results.
This integration enhances Lula's capabilities by leveraging Kyverno's policy-based validation approach. It provides flexibility in defining multiple policies, controlling validation outcomes, and observing specific rule violations.

Please refer to the provided documentation in the PR description for more details on the payload structure and usage examples.

We believe this integration will greatly benefit users who require compliance validation using Kyverno policies within the Lula framework.

### CODE REVIEW

File: docs/kyverno-provider.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+target:
+  provider: "kyverno"
+  domain: "kubernetes"
+  payload:
+    resource-rules: 
+    - group: 
+      version: v1 
+      resource: pods
+      namespaces: [validation-test] 
+    rego: |
```
Comment (brandtkeller) [inline docs/kyverno-provider.md]: Copy/Paste error here - remove the rego field in favor of the current child `kyverno` field

File: src/pkg/providers/kyverno/kyverno.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		return matchResult, nil
+	}
+
+	policies, err := policy.Parse([]byte(kyvernoPolicy))
+	if err != nil {
+		message.Debugf("Failed to parse Kyverno policy: %v", err)
+		return matchResult, fmt.Errorf("failed to parse Kyverno policy: %w", err)
+	}
+
+	var validation map[string]map[string]bool
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go]: This map is initialized if I understand correctly.

File: src/pkg/providers/kyverno/kyverno.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+			if len(pair) != 2 {
+				message.Debugf("Invalid validation pair: %v", pair)
+				continue
+			}
+
+			validation[strings.TrimSpace(pair[0])][strings.TrimSpace(pair[1])] = true
+		}
+	}
+
+	var observation map[string]map[string]bool
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go]: This map is initialized if I understand correctly.

File: src/pkg/providers/kyverno/kyverno.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+			if _, ok := validation[policy.Policy.Name][rule.Rule.Name]; output.Validation == "" || ok {
+				if len(rule.Violations) > 0 {
+					matchResult.Failing += 1
+				} else {
+					matchResult.Passing += 1
+				}
+			}
+
+			if _, ok := observation[policy.Policy.Name][rule.Rule.Name]; len(output.Observations) == 0 || ok {
+				matchResult.Observations[fmt.Sprintf("%s,%s-%d,%d", policy.Policy.Name, rule.Rule.Name, i, j)] = rule.Violations[0].Message
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go]: As noted on line 151 - in the event there are not violations - then this line will result in an `index out of range` error. Likely need a conditional for the success scenario. 

Additionally - this assignment will also incur an error with an uninitialized map. I would recommend creation of a temporary map to then assign to `matchResult.Observations` before returning.

**Review by brandtkeller — CHANGES_REQUESTED** [review]

**Comment by brandtkeller** [comment]
@KhaledEmaraDev Appreciate the direction this is driving. If you need anything - feel free to tag me here or on the Kubernetes Slack.

**Comment by KhaledEmaraDev** [comment]
Thanks @brandtkeller!

What is your opinion on the `output` field? Is the way I used it a good one: To control which policy outputs observations?

A lot of structures are OPA specific. Should we try to do a new major version and abstract theses a bit?

File: src/test/e2e/scenarios/pod-label/oscal-component-kyverno.yaml [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                kind: ValidatingPolicy
+                metadata:
+                  name: labels
+                spec:
+                  rules:
+                  - name: foo-label-exists
+                    assert:
+                      all:
+                      - check:
+                          ~.podsvt:
```
Comment (brandtkeller) [inline src/test/e2e/scenarios/pod-label/oscal-component-kyverno.yaml]:
```suggestion
                          podvt:
```

File: src/pkg/providers/kyverno/kyverno.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+		for _, pair := range validationPairs {
+			pair := strings.Split(pair, ".")
+
+			if len(pair) != 2 {
+				message.Debugf("Invalid validation pair: %v", pair)
+				continue
+			}
+
+			validation[strings.TrimSpace(pair[0])][strings.TrimSpace(pair[1])] = true
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go]: You've initialized the map on line 107 - but I believe you'll still encounter `assignment to entry in nil map` error here due to the second map `map[string]bool` also not being initialized. Couple ways to solve this - but a conditional before this assignment should suffice.

Something like: 
```
if _, ok := validation[strings.TrimSpace(pair[0])]; !ok {
    validation[strings.TrimSpace(pair[0])] = make(map[string]bool)
}
```

File: src/pkg/providers/kyverno/kyverno.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	if len(output.Observations) > 0 {
+		for _, observationPair := range output.Observations {
+			pair := strings.Split(observationPair, ".")
+
+			if len(pair) != 2 {
+				message.Debugf("Invalid validation pair: %v", pair)
+				continue
+			}
+
+			observation[strings.TrimSpace(pair[0])][strings.TrimSpace(pair[1])] = true
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go]: This will incur the same issue as above for `assignment to entry in nil map`

**Comment by brandtkeller** [comment]
> Thanks @brandtkeller!
> 
> What is your opinion on the `output` field? Is the way I used it a good one: To control which policy outputs observations?
> 
> A lot of structures are OPA specific. Should we try to do a new major version and abstract theses a bit?

@KhaledEmaraDev I believe the way you've implemented  `output.validation` is perfect and very configurable for optionality of 1 -> N combinations of policy/rule.

As for `output.observations` - I believe you're halfway there currently. The logic is present to map any targeted observations in the presence of violations. When there are no violations - given there is no other information to report - a generic observation that indicates "success" would more than suffice. 

Please call out anything you think we need to abstract - and we can consider/file issues. The project not reached GA currently - so we are expecting some breaking changes.

File: src/pkg/providers/kyverno/kyverno.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+			if _, ok := validation[policy.Policy.Name][rule.Rule.Name]; output.Validation == "" || ok {
+				if len(rule.Violations) > 0 {
+					matchResult.Failing += 1
+				} else {
+					matchResult.Passing += 1
+				}
+			}
+
+			if _, ok := observation[policy.Policy.Name][rule.Rule.Name]; len(output.Observations) == 0 || ok {
+				if len(rule.Violations) > 0 {
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go:170]: This is adding an observation when violations are present. Observations can also add information when successful.

Might be good to add an else here that adds a generic success response (unless there is other success data to map from the rule. 

This also might mean renaming 'observedViolations' to a more generic name.

**Review by brandtkeller — CHANGES_REQUESTED** [review]
Added a thought for the observations. Also requesting to sign your commits as we have identified in the contributing docs. Otherwise this looks close to ready!

**Comment by KhaledEmaraDev** [comment]
Thanks @brandtkeller, resolved.

File: docs/kyverno-provider.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                          # Check that the image tag is not `:latest`
+                          (ends_with(@, ':latest')): false
+```
+
+You can have mutiple policies defined. Optionally, `output.validation` can be specified in the `payload` to control which (Policy, Rule) pair control validation allowance/denial, which is in the structure of a comma separated list of rules: `policy-name1.rule-name-1,policy-name-1.rule-name-2`. If you have a desired observation to include, `output.observations` can be added to payload to observe violations by a certain (Policy, Rule) pair such as:
+```yaml
+target:
+  provider: "kyverno"
+  domain: "kubernetes"
+  payload:
```
Comment (brandtkeller) [inline docs/kyverno-provider.md:56]: Missing the `resources` field with child `resource-rule`

File: src/pkg/providers/kyverno/kyverno.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+					matchResult.Failing += 1
+				} else {
+					matchResult.Passing += 1
+				}
+			}
+
+			if _, ok := observationSet[policy.Policy.Name][rule.Rule.Name]; len(output.Observations) == 0 || ok {
+				if len(rule.Violations) > 0 {
+					observations[fmt.Sprintf("%s,%s-%d,%d", policy.Policy.Name, rule.Rule.Name, i, j)] = rule.Violations[0].Message
```
Comment (brandtkeller) [inline src/pkg/providers/kyverno/kyverno.go]: On a failing scenario where `message` was not specified in the rule - `rule.Violations[0].Message` will be an empty string I believe. 

Example failing scenario (I made foo-label-single-exists` failing
```
"labels,foo-label-exists-0,0: PASS\nlabels,foo-label-single-exists-0,1: \n"
```

Might be good practice to prepend a `FAIL` to have at a minimum a default message when the user did not include the `message` field on a rule. 

```
observations[fmt.Sprintf("%s,%s-%d,%d", policy.Policy.Name, rule.Rule.Name, i, j)] = "FAIL " + rule.Violations[0].Message
```

(or similar)

**Review by mike-winberry — DISMISSED** [review]
Looks good, pending @brandtkeller requested changes.

**Review by brandtkeller — APPROVED** [review]
Tested a variety of permutations to work through testing the Kyverno provider. 

To include:
```
target:
          provider: kyverno
          domain: kubernetes
          payload:
            resources:
            - name: podsvt                      # Identifier for use in the rego below
              resource-rule:                     # Mandatory, resource selection criteria, at least one resource rule is required
                group:                          # empty or "" for core group
                version: v1                     # Version of resource
                resource: pods                  # Resource type
                namespaces: [validation-test]   # Namespaces to validate the above resources in. Empty or "" for all namespaces or non-namespaced resources
            - name: podvt                      # Identifier for use in the rego below
              resource-rule:                     # Mandatory, resource selection criteria, at least one resource rule is required
                name: demo-pod
                group:                          # empty or "" for core group
                version: v1                     # Version of resource
                resource: pods                  # Resource type
                namespaces: [validation-test]
            kyverno: |
              apiVersion: json.kyverno.io/v1alpha1
              kind: ValidatingPolicy
              metadata:
                name: labels
              spec:
                rules:
                - name: foo-label-exists
                  assert:
                    all:
                    - check:
                        ~.podsvt:
                          metadata:
                            labels:
                              foo: bar
                - name: foo-label-single-exists
                  assert:
                    all:
                    - message: all pods need to have label foo=bar
                      check:
                        podvt:
                          metadata:
                            labels:
                              foo: bar
            output:
              validation: labels.foo-label-single-exists,labels.foo-label-exists
              observations:
              - labels.foo-label-exists
              - labels.foo-label-single-exists
```

This all looks great!

**Review by mike-winberry — APPROVED** [review]

**Pull request merged**